src/node/stage/approved.rs

```diff
@@ -678,225 +729,268 @@ impl Approved {
-                        "parsec_poll polled internal Observation {}: {:?}",
-                        parsec_version,
-                        obs
-                    );
-                }
-                Observation::DkgResult {
-                    participants,
-                    dkg_result,
-                } => {
-                    self.chain
-                        .handle_dkg_result_event(participants, dkg_result)?;
-                    self.handle_dkg_result_event(participants, dkg_result)?;
-                }
-            }
+        // Note: it's important that `promote_and_demote_elders` happens before `poll_relocation`,
+        // otherwise we might relocate a node that we still need.
+        if self.promote_and_demote_elders(core)? {
+            return Ok(true);
+        }
 
-            self.chain_poll(core)?;
+        if self.poll_relocation(core.id()) {
+            return Ok(true);
         }
 
-        self.check_voting_status();
+        let (event, proof) = match self.consensus_engine.poll(self.shared_state.sections.our()) {
+            None => return Ok(false),
+            Some((event, proof)) => (event, proof),
+        };
 
-        Ok(())
+        if let Some(event) = self.check_ready_or_backlog_churn_event(event) {
+            self.handle_accumulated_event(core, event, proof)?;
+            return Ok(true);
+        }
+
+        Ok(false)
     }
 
-    fn chain_poll(&mut self, core: &mut Core) -> Result<()> {
-        let mut old_pfx = *self.chain.our_prefix();
-        let mut was_elder = self.chain.is_self_elder();
+    pub fn can_poll_churn(&self) -> bool {
+        self.shared_state.handled_genesis_event && !self.churn_in_progress
+    }
 
-        while let Some(event) = self.chain.poll_accumulated()? {
-            match event {
-                PollAccumulated::AccumulatedEvent(event) => {
-                    self.handle_accumulated_event(core, event, old_pfx, was_elder)?
-                }
-                PollAccumulated::RelocateDetails(details) => {
-                    self.handle_relocate_polled(details)?;
-                }
-                PollAccumulated::PromoteDemoteElders(new_infos) => {
-                    self.handle_promote_and_demote_elders(new_infos)?;
-                }
+    // Polls and processes a backlogged churn event, if any.
+    fn poll_churn_event_backlog(&mut self, core: &mut Core) -> Result<bool> {
+        if !self.can_poll_churn() {
+            return Ok(false);
+        }
+
+        if let Some(event) = self.shared_state.churn_event_backlog.pop_back() {
+            trace!(
+                "churn backlog poll {:?}, Others: {:?}",
+                event,
+                self.shared_state.churn_event_backlog
+            );
+
+            self.handle_accumulated_event(core, event, AccumulatingProof::default())?;
+            Ok(true)
+        } else {
+            Ok(false)
+        }
+    }
+
+    fn check_ready_or_backlog_churn_event(
```

Comment:
And this is supposed to mean: can we process this event right now or do we need to put it into the backlog because we are in the middle of churn? Better naming suggestions welcome, but note that this function doesn't just return `true`/`false`. It either returns the event which means we can process it, or it backlogs it and returns `None`.

Reply:
aha ok, that whole thing seems a bit convoluted..

I would consider it better to keep the query part separated, so we get this:

in `fn poll_one()`:

```
let poll_one = if self.shall_backlog_event(event) {
            self.shared_state.churn_event_backlog.push_front(event);
            false
        } else {
            self.handle_accumulated_event(core, event, proof)?;
            true
        };

        Ok(poll_one)
```

and `check_ready_or_backlog_churn_event` is replaced with

```
fn shall_backlog_event(&mut self, event: AccumulatingEvent) -> bool {
        let is_churn_trigger = match &event {
            AccumulatingEvent::Online(_)
            | AccumulatingEvent::Offline(_)
            | AccumulatingEvent::Relocate(_) => true,
            _ => false,
        };
        is_churn_trigger && !self.can_poll_churn()
    }
```

More straight forward IMO, clearer.

Reply:
Sounds good. Will do.